File: haskell/binary-search/src/BinarySearch.hs
module BinarySearch (find) where

import Data.Array

binSearch :: Ord a => Array Int a -> a -> Int -> Int -> Maybe Int
binSearch array target start end
    | start > end = Nothing
    | middleValue == target = Just middle
    | middleValue > target = binSearch array target start (middle-1)
    | otherwise = binSearch array target (middle+1) end
    where middle = (end + start) `div` 2
          middleValue = array ! middle

find :: Ord a => Array Int a -> a -> Maybe Int
find array x = binSearch array x start end
  where
  (start, end) = bounds array
  

File: scala/binary-search/src/main/scala/BinarySearch.scala
object BinarySearch {
  def find(songs: List[Int], favoriteNumber: Int): Option[Int] = {
    def  binSearch(lo: Int, hi: Int): Option[Int] = {
      if(hi < lo) return None
      val index = (lo+hi)/2
      if(songs(index) == favoriteNumber) Some(index)
      else if(songs(index) < favoriteNumber) binSearch(index+1, hi)
      else binSearch(lo, index-1)
    }
    binSearch(0, songs.length-1)
  }
}